components/style/values/generics/image.rs

```diff
@@ -0,0 +1,637 @@
+    /// An ellipse pair of radii.
+    Radii(LengthOrPercentage, LengthOrPercentage),
+    /// An ellipse extent.
+    Extent(ShapeExtent),
+}
+
+/// https://drafts.csswg.org/css-images/#typedef-extent-keyword
+define_css_keyword_enum!(ShapeExtent:
+    "closest-side" => ClosestSide,
+    "farthest-side" => FarthestSide,
+    "closest-corner" => ClosestCorner,
+    "farthest-corner" => FarthestCorner,
+    "contain" => Contain,
+    "cover" => Cover
+);
+
+/// A gradient item.
+/// https://drafts.csswg.org/css-images-4/#color-stop-syntax
+#[derive(Clone, Copy, PartialEq, Debug)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub enum GradientItem<Color, LengthOrPercentage> {
+    /// A color stop.
+    ColorStop(ColorStop<Color, LengthOrPercentage>),
+    /// An interpolation hint.
+    InterpolationHint(LengthOrPercentage),
+}
+
+/// A color stop.
+/// https://drafts.csswg.org/css-images/#typedef-color-stop-list
+#[derive(Clone, Copy, PartialEq)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub struct ColorStop<Color, LengthOrPercentage> {
+    /// The color of this stop.
+    pub color: Color,
+    /// The position of this stop.
+    pub position: Option<LengthOrPercentage>,
+}
+
+/// Values for `moz-image-rect`.
+///
+/// `-moz-image-rect(<uri>, top, right, bottom, left);`
+#[derive(Clone, PartialEq, Debug)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+#[allow(missing_docs)]
+pub struct ImageRect<NumberOrPercentage> {
+    pub url: SpecifiedUrl,
+    pub top: NumberOrPercentage,
+    pub bottom: NumberOrPercentage,
+    pub right: NumberOrPercentage,
+    pub left: NumberOrPercentage,
+}
+
+impl<G, R> fmt::Debug for Image<G, R>
+    where G: fmt::Debug, R: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Image::Url(ref url) => url.to_css(f),
+            Image::Gradient(ref grad) => grad.fmt(f),
+            Image::Rect(ref rect) => rect.fmt(f),
+            Image::Element(ref selector) => {
+                f.write_str("-moz-element(#")?;
+                serialize_identifier(&selector.to_string(), f)?;
+                f.write_str(")")
+            },
+        }
+    }
+}
+
+impl<G, R> ToCss for Image<G, R>
+    where G: ToCss, R: ToCss,
+{
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        match *self {
+            Image::Url(ref url) => url.to_css(dest),
+            Image::Gradient(ref gradient) => gradient.to_css(dest),
+            Image::Rect(ref rect) => rect.to_css(dest),
+            Image::Element(ref selector) => {
+                dest.write_str("-moz-element(#")?;
+                serialize_identifier(&selector.to_string(), dest)?;
+                dest.write_str(")")
+            },
+        }
+    }
+}
+
+impl<G, R> HasViewportPercentage for Image<G, R>
+    where G: HasViewportPercentage
+{
+    fn has_viewport_percentage(&self) -> bool {
+        if let Image::Gradient(ref gradient) = *self {
```

Comment:
nit: I think `match` is nicer than `if let { ... } else { }`, but feel free to disagree on this :)